Pegasus/include/pegasus/Collision.hpp

```diff
@@ -0,0 +1,250 @@
+                        std::ref(m_assetManager->GetAsset(m_assetManager->GetBodies(), bObject.data.body)),
+                        CalculateContactManifold(aShape, bShape),
+                        0.75
+                    );
+                    registeredContacts.insert(key);
+                }
+            }
+        }
+
+        return contacts;
+    }
+
+    /**
+     * @brief Calculates contacts between two sets of rigid bodies
+     * @tparam ObjectA rigid body type
+     * @tparam ShapeA shape type
+     * @tparam ObjectB rigid body type
+     * @tparam ShapeB shape type
+     * @return contacts
+     */
+    template < typename ObjectA, typename ShapeA, typename ObjectB, typename ShapeB >
+    std::vector<Contact> Detect()
+    {
+        std::vector<Contact> contacts;
+        std::unordered_set<std::pair<ShapeA*, ShapeB*>, ObjectHasher> registeredContacts;
+        std::vector<scene::Asset<scene::RigidBody>>& aObjects = m_assetManager->GetObjects<ObjectA, ShapeA>();
+        std::vector<scene::Asset<scene::RigidBody>>& bObjects = m_assetManager->GetObjects<ObjectB, ShapeB>();
+
+        for (scene::Asset<scene::RigidBody> aObject : aObjects)
+        {
+            for (scene::Asset<scene::RigidBody> bObject : bObjects)
+            {
+                if (aObject.id == 0 || bObject.id == 0)
+                    continue;
+
+                ShapeA* aShape = &m_assetManager->GetAsset(m_assetManager->GetShapes<ShapeA>(), aObject.data.shape);
+                ShapeB* bShape = &m_assetManager->GetAsset(m_assetManager->GetShapes<ShapeB>(), bObject.data.shape);
+                std::pair<ShapeA*, ShapeB*> const key = std::make_pair(aShape, bShape);
+
+                if (Intersect(aShape, bShape)
+                    && registeredContacts.find(key) == registeredContacts.end())
+                {
+                    contacts.emplace_back(
+                        std::ref(m_assetManager->GetAsset(m_assetManager->GetBodies(), aObject.data.body)),
+                        std::ref(m_assetManager->GetAsset(m_assetManager->GetBodies(), bObject.data.body)),
+                        CalculateContactManifold(aShape, bShape),
+                        0.75
+                    );
+                    registeredContacts.insert(key);
+                }
+            }
+        }
+
+        return contacts;
+    }
+};
+
+class Resolver
+{
+public:
+    uint32_t iterationsUsed;
+    uint32_t iterations = 10000;
+
+    /**
+     * @brief Resolves collisions
+     * @param contacts contacts information
+     * @param duration delta time of the frame
+     */
+    void Resolve(std::vector<std::vector<Contact>>& contacts, double duration);
+
+private:
+    /**
+     * @brief Calculates total separation speed of the contact
+     * @param contact contact information
+     * @return speed meters per second
+     */
+    static double CalculateTotalSeparationSpeed(Contact contact);
```

Comment:
In future, it might help in parallelization, because every thread would have a local copy of a contact object, which allows for more non-concurrent and non-blocking solutions. Also, it improves locality of the data. However, in future, if proven not useful it might be changed.

Reply:
Okay.